src/main/java/org/broadinstitute/hellbender/tools/spark/bwa/BwaSparkEngine.java

```diff
@@ -129,8 +133,17 @@ public void close() {
                 inputReads.add(readItr.next());
             }
             final int nReads = inputReads.size();
-            if (alignsPairs && (nReads & 1) != 0 ) {
-                throw new GATKException("We're supposed to be aligning paired reads, but there are an odd number of them.");
+            if ( alignsPairs ) {
+                if ( (nReads & 1) != 0 ) {
+                    throw new GATKException("We're supposed to be aligning paired reads, but there are an odd number of them.");
+                }
+                for ( int idx = 0; idx != nReads; idx += 2 ) {
+                    final String readName1 = inputReads.get(idx).getName();
+                    final String readName2 = inputReads.get(idx+1).getName();
+                    if ( !Objects.equals(readName1,readName2) ) {
+                        throw new GATKException("Read pair has varying template name: "+readName1+" .vs "+readName2);
```

Comment:
Should this one be a `UserException`, since it's a problem with the input?

Reply:
fixed